enclave/crypto/openssl/cert.c

```diff
@@ -0,0 +1,457 @@
+#include <openssl/x509_vfy.h>
+#include <openssl/x509v3.h>
+#include <string.h>
+
+/*
+ * Value used by the _decode_oid_to_str() function. Although the OID
+ * standard does not limit the depth of an OID definition tree (i.e., the
+ * number of arcs), our implementation only supports a simple decoding
+ * with a limited depth (i.e., decoding into a fixed size string).
+ */
+#define OE_OID_MAX_LENGTH 200
+
+#define OE_X509_MAX_NAME_SIZE 256
+
+/*
+ * Parse the name string into X509_NAME struct. The format of the string is
+ * "KEY1=VALUE1,KEY2=VALUE2,KEY3=VALUE3...". The implementation is based
+ * on the mbedtls_x509_string_to_names from Mbed TLS.
+ */
+X509_NAME* X509_parse_name(const char* name_string)
+{
+    const char* s = name_string;
+    const char* c = s;
+    const char* end = s + strlen(s);
+    int in_tag = 1;
+    char key[OE_X509_MAX_NAME_SIZE];
+    char value[OE_X509_MAX_NAME_SIZE];
+    char data[OE_X509_MAX_NAME_SIZE];
+    char* d = data;
+    X509_NAME* name = NULL;
+    int error = 1;
+
+    name = X509_NAME_new();
+    if (name == NULL)
+        goto done;
+
+    while (c <= end)
+    {
+        if (in_tag && *c == '=')
+        {
+            size_t len = (size_t)(c - s) + 1;
+            if (len > OE_X509_MAX_NAME_SIZE)
+                goto done;
+
+            if (oe_memcpy_s(key, OE_X509_MAX_NAME_SIZE, s, len) != OE_OK)
+                goto done;
+            key[len - 1] = '\0';
+            s = c + 1;
+            in_tag = 0;
+            d = data;
+        }
+
+        if (!in_tag && *c == '\\' && c != end)
+        {
+            c++;
+            /* Only support escaping commas */
+            if (c == end || *c != ',')
+                goto done;
+        }
+        else if (!in_tag && (*c == ',' || c == end))
+        {
+            /*
+             * The check of if(d - data == OE_X509_MAX_NAME_SIZE)
+             * below ensures that len should never exceed OE_X509_MAX_NAME_SIZE.
+             * We do not perform additional checks here.
+             */
+            size_t len = (size_t)(d - data) + 1;
+            if (oe_memcpy_s(value, OE_X509_MAX_NAME_SIZE, data, len) != OE_OK)
+                goto done;
+            value[len - 1] = '\0';
+            if (!X509_NAME_add_entry_by_txt(
+                    name, key, MBSTRING_ASC, (unsigned char*)value, -1, -1, 0))
+                goto done;
+
+            /* Skip the spaces after the comma */
+            while (c < end && *(c + 1) == ' ')
+                c++;
+            s = c + 1;
+            in_tag = 1;
+        }
+
+        if (!in_tag && s != c + 1)
+        {
+            *(d++) = *c;
+            if (d - data == OE_X509_MAX_NAME_SIZE)
+                goto done;
+        }
+
+        c++;
+    }
+
+    error = 0;
+
+done:
+    if (error && name)
+    {
+        X509_NAME_free(name);
+        name = NULL;
+    }
+
+    return name;
+}
+
+/*
+ * Decode the OID from its BER format (a hex string) into a dot-notation string
```

Comment:
I would suggest referring to this as a _byte-encoding_ rather than a hex string as the latter can be misunderstood as "0xA2" rather than what you intend here.

Reply:
Updated.